### PR TITLE
Login

### PR DESCRIPTION
**Pull request by ahow6552**
# Login

## Description

Complete login/out function.

## Checklist

| Item          | Is checked? | Description                                        |
| ------------- | ----------- | -------------------------------------------------- |
| Explanation   | X           | My changes are well-explained.                     |
| Documentation | X           | My works are well-commented or clear.              |
| Tests         | X           | I have added tests or reports to cover my changes. |

### CODE REVIEW

**Comment by ahow6552** [comment]
@AsherJingkongChen

**Comment by AsherJingkongChen** [comment]
@ahow6552 Remember to complete your description, or I can complete it for you.

**Review by AsherJingkongChen — CHANGES_REQUESTED** [review]
## Blocking

Requires to solve the conflicts

File: app/routes.py [hunk 1]
```diff
@@ -20,27 +20,52 @@
...
-def get_data():
+@bp_web_api.route('/auth/login', methods=['GET', 'POST'])  
+def login():   
+    if request.method == 'GET':  
+           return '''
+     <form action='login' method='POST'>
+     <input type='text' name='name' id='name' placeholder='name'/>
+     <input type='password' name='credential' id='credential' placeholder='credential'/>
+     <input type='submit' name='submit'/>
+     </form>
```
Comment (AsherJingkongChen) [inline app/routes.py]: ## Warn

### Description

Redundant and undefined in Web API Spec

### Solution

```suggestion
```

File: app/routes.py [hunk 2]
```diff
@@ -20,27 +20,52 @@
...
+                  '''
+
     conn = sqlite3.connect('data/nbaDB.db')
     cursor = conn.cursor()
-    cursor.execute('SELECT * FROM Team;')
+    cursor.execute('SELECT * FROM Manager;')
```
Comment (AsherJingkongChen) [inline app/routes.py]: ## Bad

Use `WHERE` clause to filter

File: app/routes.py [hunk 3]
```diff
@@ -20,27 +20,52 @@
...
-        data.append(
-            {
-                'column1': row[0],
-                'column2': row[1],
-            }
-        )
-    return jsonify(data)
+    data = request.get_json()
+
+    name = data.get('name')
```
Comment (AsherJingkongChen) [inline app/routes.py]: ## Warn

注意 Web API Spec `name`, `credential` 為必填欄位

File: app/routes.py [hunk 4]
```diff
@@ -20,27 +20,52 @@
...
+    session_id = request.cookies.get('session_id')  
+    if session_id :
+        response = Response(status=204)
+        cookie['session_id'] = ''
+        response.set_cookie('session_id', '', httponly=True, max_age=0, path='/', samesite='Strict')
+        return response  
+    
+    return jsonify({"message": "You are not authorized to access this resource."}), 404
 
 @bp_web_api.route('/auth/login', methods=['POST'])
```
Comment (AsherJingkongChen) [inline app/routes.py:69]: ## Warn

Dupe endpoints

File: app/routes.py [hunk 5]
```diff
@@ -20,27 +20,52 @@
...
+            response = Response(status=201)
+            response.set_cookie('session_id', session_id, httponly=True, max_age=31536000, path='/', samesite='Strict')
+        return response
+    
+    return jsonify({"message": "The resource you are accessing is not found."}), 404
+
+@bp_web_api.route('/auth/logout', methods=['GET'])  
+def logout():
+    session_id = request.cookies.get('session_id')  
+    if session_id :
```
Comment (AsherJingkongChen) [inline app/routes.py]: ## Warn

Bad condition

**Review by AsherJingkongChen — CHANGES_REQUESTED** [review]
## Blocking

@ahow6552 Are you making stubs or real implementations?

There are several warnings.

File: app/routes.py [hunk 6]
```diff
@@ -20,27 +20,52 @@
...
 
+cookie = {}
+
 @bp_web_api.route('/ping', methods=['GET'])
 def ping():
     return jsonify('Pong!')
 
-@bp_web_api.route('/data', methods=['GET'])
-def get_data():
+@bp_web_api.route('/auth/login', methods=['GET', 'POST'])
```
Comment (AsherJingkongChen) [inline app/routes.py]: ## Warn

No GET is defined for `/auth/login`

```suggestion
@bp_web_api.route('/auth/login', methods=['POST'])
```

File: app/routes.py [hunk 7]
```diff
@@ -20,27 +20,52 @@
...
+        if data.get('credential') == password and name == username :  
+            session_id = str(uuid.uuid4().hex)
+            cookie['session_id'] = session_id
+            response = Response(status=201)
+            response.set_cookie('session_id', session_id, httponly=True, max_age=31536000, path='/', samesite='Strict')
+        return response
+    
+    return jsonify({"message": "The resource you are accessing is not found."}), 404
+
+@bp_web_api.route('/auth/logout', methods=['GET'])
```
Comment (AsherJingkongChen) [inline app/routes.py]: ## Error

Wrong method

### Solution

```suggestion
@bp_web_api.route('/auth/logout', methods=['POST'])
```

File: app/routes.py [hunk 8]
```diff
@@ -20,27 +20,52 @@
...
-            }
-        )
-    return jsonify(data)
+    data = request.get_json()
+
+    name = data.get('name')
+    for username, password in rows:
+        if data.get('credential') == password and name == username :  
+            session_id = str(uuid.uuid4().hex)
+            cookie['session_id'] = session_id
```
Comment (AsherJingkongChen) [inline app/routes.py]: ## Warn

State 預設是允許多 session, implied by semantics

File: app/routes.py [hunk 9]
```diff
@@ -20,27 +20,52 @@
...
+            response = Response(status=201)
+            response.set_cookie('session_id', session_id, httponly=True, max_age=31536000, path='/', samesite='Strict')
+        return response
```
Comment (AsherJingkongChen) [inline app/routes.py]: ## Error

Unhandled Exception:

```python
UnboundLocalError: cannot access local variable 'response' where it is not associated with a value
```

### Solution

Check your return values
```suggestion
            return response
```

**Review by AsherJingkongChen — CHANGES_REQUESTED** [review]

File: app/routes.py (hunk 3 above)
Comment (ahow6552) [reply app/routes.py]: @AsherJingkongChen 這沒填不就是空的嗎

Comment (AsherJingkongChen) [reply app/routes.py]: @ahow6552

你是要確認 API Spec 的合理性？

那麼空使用者的定義是什麼？空密碼也不是正常狀況。

如果對 API Spec 有問題應該到 doc/issue 提問題。

File: app/routes.py [hunk 10]
```diff
@@ -20,27 +20,52 @@
     template_folder='../web/dist',
 )
 
+cookie = {}
```
Comment (AsherJingkongChen) [inline app/routes.py]: ## Suggestion

Give this variable a comment

**Review by AsherJingkongChen — CHANGES_REQUESTED** [review]

**Pull request closed without merging**